rfc/001-composable-pipelines.md

```diff
@@ -0,0 +1,205 @@
+# Composable Pipelines
+
+- Authors:
+  - [Matthias Vallentin](https://github.com/mavam)
+
+## Overview
+
+This proposal ideates a new UX around VAST's pipelines. The fundamental idea is
+that *everything is a pipeline* and build a composable UX around this concept.
+
+## Problem Statement
+
+Pipelines currently are only deployable at [predefined
+points](https://vast.io/docs/use-vast/transform). For example, they can be
+triggered for specific events on import/export, or invoked by compaction.
+
+As we are bringing pipelines to the query language, we need to solve a UX
+challenge: how do we make the data flow explicit and provide a unified
+experience, especially in the presence of custom plugins?
+
+## Solution Proposal
+
+Assume for a moment everything is a pipeline operator. Then we may write:
+
+ ```bash
+vast 'from s3://aws |
+      read json |
+      group-by src |
+      summarize count(dst) |
+      write feather |
+      to /path/to/file.feather'
+```
+
+Just to illustrate the composability, we could take this to an extreme when
+performance doesn't matter:
+
+```bash
+vast from s3://aws |
+  vast read json |
+  vast group-by |
+  vast summarize |
+  vast write feather |
+  vast to /path/to/file.feather
+```
+
+But let's take a step back and just assume that we add a new `exec` command that
+just executes a pipeline.
+
+Then we can model the ingestion as follows: load data via stdin, push into a
+parser, then send off to a remote VAST node. For example, we would rewrite
```

Comment:
Yeah, this hinges on what we are going to do with exact duplicates. Let's say we had a deduplication toggle, then this would be a no-op and refeed the entire DB into itself:

```bash
vast exec 'from vast://1.2.3.4  | to vast://1.2.3.4
```

In your example, you add a column. So deduplication wouldn't kick in. So VAST would consider the augmented records as new data.

(NB: the `add` operator is what we call `extend`, akin to Kusto.)

Reply:
Makes sense, I guess this thread is then converging to the thread about having mutating pipelines (in addition to non-mutable) https://github.com/tenzir/vast/pull/2511#discussion_r950010277

(Feel free to resolve the thread)